### PR TITLE
Move `Headers` into axum

### PR DESCRIPTION
**Pull request by davidpdrsn**
- Moves `Headers` from axum-core and into axum.
- Removes the blanket `impl<H> IntoResponse for H where H: IntoResponseHeaders`. This really hurt the error message from `#[debug_handler]` as it would mention the `IntoResponseHeaders` trait instead of `IntoResponse`.
- Implement `IntoResponse` for `HeaderMap` and `Headers`. Having to do this is a bit unfortunate but I'd rather have the better error message for `#[debug_handler]`.
- This also means `IntoResponseHeaders` can be implemented outside of axum-core. I guess thats fine 🤷

### CODE REVIEW

File: axum-core/src/response/mod.rs [hunk 1]
```diff
@@ -149,38 +144,17 @@ pub trait IntoResponse {
...
-///
-/// ```text
-/// note: required because of the requirements on the impl of `IntoResponse` for `<type>`
-/// ```
-///
-/// Any type that implements this trait automatically implements `IntoResponse` as well, but can
-/// also be used in a tuple like `(StatusCode, Self)`, `(Self, impl IntoResponseHeaders)`,
-/// `(StatusCode, Self, impl IntoResponseHeaders, impl IntoResponse)` and so on.
-///
-/// This trait can't currently be implemented outside of axum.
```
Comment (jplatte) [inline axum-core/src/response/mod.rs:166]: The current trait members are rather weird, do you really want them to be part of the public API nominally?

File: axum-core/src/response/mod.rs [hunk 2]
```diff
@@ -356,17 +330,10 @@ impl IntoResponse for StatusCode {
     }
 }
 
-impl<H> IntoResponse for H
-where
-    H: IntoResponseHeaders,
-{
+impl IntoResponse for HeaderMap {
```
Comment (jplatte) [inline axum-core/src/response/mod.rs:333]: I guess this makes sense. It's really annoying the previous solution worsens rustc's error reporting.

File: axum-core/src/response/mod.rs (hunk 1 above)
Comment (davidpdrsn) [reply axum-core/src/response/mod.rs:166]: Yeah they're quite odd at first glance. I suppose we could make them `#[doc(hidden)]` 🤔

File: axum-core/src/response/mod.rs (hunk 2 above)
Comment (davidpdrsn) [reply axum-core/src/response/mod.rs:333]: Yeah it is 😞 I still think its worth it overall

Comment (davidpdrsn) [reply axum-core/src/response/mod.rs:333]: (including the introduction of `IntoResponseHeaders`)

Comment (jplatte) [reply axum-core/src/response/mod.rs:333]: Yeah, agreed.

File: axum-core/src/response/mod.rs (hunk 1 above)
Comment (jplatte) [reply axum-core/src/response/mod.rs:166]: I think we should.

**Review by jplatte — APPROVED** [review]
LGTM with `#[doc(hidden)]` re-added.

File: axum-core/src/response/mod.rs [hunk 3]
```diff
@@ -149,37 +144,18 @@ pub trait IntoResponse {
...
     ///
-    /// The header name is optional because `HeaderMap`s iterator doesn't yield it multiple times
+    /// The header name is optional because [`HeaderMap`]s iterator doesn't yield it multiple times
     /// for headers that have multiple values, to avoid unnecessary copies.
     #[doc(hidden)]
     type IntoIter: IntoIterator<Item = Result<(Option<HeaderName>, HeaderValue), Response>>;
 
     /// Attempt to turn `self` into a list of headers.
-    ///
-    /// In practice, only the implementation for `axum::response::Headers` ever returns `Err(_)`.
```
Comment (jplatte) [inline axum-core/src/response/mod.rs:182]: Why was this deleted, by the way?

Comment (davidpdrsn) [reply axum-core/src/response/mod.rs:182]: Ups not sure. Added it back in 208720be531df.
